File: util/setup/editor.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

// The extension is linked, never copied. The build renders the tree, and this
// points the editor at it, so a rebuild needs no second install.
//
// BUILDING IT AND LINKING IT ARE TWO THINGS, AND ONLY THE SECOND NEEDS AN
// EDITOR. Both sat in one function that ran on the desktop profile alone, so a
// headless box never ran npm install and src/extension/node_modules was never
// made. Four checks bundle the extension's TypeScript with the esbuild in that
// folder, and on a fresh Linux clone all four failed with ERR_MODULE_NOT_FOUND
// before they read a line of the tree. The dependencies are the battery's, so
// every profile installs them; the link into the editor's extensions folder
// needs an editor, so it stays behind the desktop profile.

func extensionsDir() string { return filepath.Join(homeDir(), ".vscode", "extensions") }

// buildExtension makes the extension's dependencies and renders it. Every
// profile runs this, because the checks read what it makes.
func buildExtension(root string) error {
	src := filepath.Join(root, "src", "extension")
	if *dry {
		say("  build    the extension would be built in %s", src)
		return nil
	}
	say("  build    the extension")
	if err := runIn(src, "npm", "install", "--silent", "--no-fund", "--no-audit"); err != nil {
		return fmt.Errorf("installing the extension's dependencies failed: %w", err)
	}
	if err := runIn(src, "npm", "run", "build"); err != nil {
		return fmt.Errorf("building the extension failed: %w", err)
	}
	return nil
}

// linkExtension points the editor at the folder buildExtension rendered.
func linkExtension(root, id string) error {
	src := filepath.Join(root, "src", "extension")
	if *dry {
		say("  editor   the extension would be linked from %s", src)
		return nil
	}
	dest := filepath.Join(extensionsDir(), id+"."+id+"-0.1.0")
	if err := os.MkdirAll(extensionsDir(), 0o755); err != nil {
		return err
	}
	if err := replaceLink(dest, src); err != nil {
		return err
	}
	say("  editor   linked at %s", dest)

	if err := registerWithEditor(dest, id); err != nil {
		return fmt.Errorf("registering the extension with the editor failed: %w", err)
	}
	say("  editor   registered")
	return nil
}

// A loaded extension cannot be replaced, because the editor holds the files it
// is running. The raw failure names an internal file and hides the cause, so
// the cause is said instead.
func replaceLink(dest, src string) error {
	if err := removeLink(dest); err != nil {
		return lockedMessage(dest, err)
	}
	if err := makeLink(dest, src); err != nil {
		return err
	}
	// Linking can report success and leave nothing usable behind. The editor
	// loads what is at the other end, so check that end.
	if _, err := os.Stat(filepath.Join(dest, "package.json")); err != nil {
		return fmt.Errorf("the link was made and the editor would find nothing through it: %s", dest)
	}
	return nil
}

func lockedMessage(dest string, err error) error {
	return fmt.Errorf("the installed extension is in use: close every editor window and run this again\n  %s\n  %v", dest, err)
}

// The editor loads what its own list names, and it does not find a linked
// folder on its own. That list holds every extension the person has, so a
// careless write loses somebody else's work. This backs the file up, keeps
// every entry it cannot read, and refuses to write if an id would be lost.
type extEntry struct {
	Identifier struct {
		ID   string `json:"id"`
		UUID string `json:"uuid,omitempty"`
	} `json:"identifier"`
	Version          string          `json:"version"`
	Location         json.RawMessage `json:"location,omitempty"`
	RelativeLocation string          `json:"relativeLocation,omitempty"`
	Metadata         json.RawMessage `json:"metadata,omitempty"`
	Rest             json.RawMessage `json:"-"`
}

func registerWithEditor(dest, id string) error {
	listPath := filepath.Join(extensionsDir(), "extensions.json")
	raw, err := os.ReadFile(listPath)
	if err != nil && !os.IsNotExist(err) {
		return err
	}

	var entries []json.RawMessage
	if len(raw) > 0 {
		if err := json.Unmarshal(raw, &entries); err != nil {
			return fmt.Errorf("the editor's extension list is not readable, so it was left alone: %w", err)
		}
	}

	before := idsIn(entries)
	fullID := id + "." + id

	kept := entries[:0:0]
	for _, e := range entries {
		if idOf(e) == fullID {
			continue // ours, replaced below
		}
		kept = append(kept, e)
	}

	uri := locationPath(dest)
	mine := map[string]any{
		"identifier":       map[string]any{"id": fullID},
		"version":          "0.1.0",
		"location":         map[string]any{"$mid": 1, "path": uri, "scheme": "file"},
		"relativeLocation": filepath.Base(dest),
		"metadata": map[string]any{
			"installedTimestamp": time.Now().UnixMilli(),
			"source":             "vsix",
		},
	}
	mineRaw, err := json.Marshal(mine)
	if err != nil {
		return err
	}
	kept = append(kept, mineRaw)

	// Nothing that was there may vanish. If it would, write nothing.
	after := idsIn(kept)
	for wasID := range before {
		if wasID == fullID {
			continue
		}
		if !after[wasID] {
			return fmt.Errorf("writing the list would have lost %s, so nothing was written", wasID)
		}
	}

	if len(raw) > 0 {
		backup := listPath + ".before-quackitect"
		if err := os.WriteFile(backup, raw, 0o644); err != nil {
			return fmt.Errorf("could not back the list up, so it was left alone: %w", err)
		}
	}
	out, err := json.Marshal(kept)
	if err != nil {
		return err
	}
	tmp := listPath + ".tmp"
	if err := os.WriteFile(tmp, out, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, listPath)
}

// The editor writes its own entries with a lower-case drive letter, because
// that is how it normalises a file URI. An entry that does not match the
// others is an entry it may not match to the folder on disk.
func locationPath(dest string) string {
	uri := filepath.ToSlash(dest)
	if runtime.GOOS == "windows" {
		if len(uri) > 1 && uri[1] == ':' {
			uri = strings.ToLower(uri[:1]) + uri[1:]
		}
		uri = "/" + uri
	}
	return uri
}

func idsIn(entries []json.RawMessage) map[string]bool {
	out := map[string]bool{}
	for _, e := range entries {
		if id := idOf(e); id != "" {
			out[id] = true
		}
	}
	return out
}

func idOf(e json.RawMessage) string {
	var probe struct {
		Identifier struct {
			ID string `json:"id"`
		} `json:"identifier"`
	}
	if json.Unmarshal(e, &probe) != nil {
		return ""
	}
	return probe.Identifier.ID
}

func runIn(dir, name string, args ...string) error {
	if _, err := exec.LookPath(name); err != nil {
		return fmt.Errorf("%s is not on this machine", name)
	}
	cmd := Quietly(exec.Command(name, args...))
	cmd.Dir = dir
	cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
	return cmd.Run()
}

func runtimeIsWindows() bool { return runtime.GOOS == "windows" }

File: src/scripts/editor.js
// THE EDITOR LOADS WHAT ITS OWN LIST NAMES, AND A LINKED FOLDER IS NOT ON IT.
// So the link into ~/.vscode/extensions goes beside an entry in that folder's
// extensions.json. That file holds every extension a person has, and v3 and v4
// both learned what a careless write costs: one unreadable element and the
// editor drops the lot, so a person watches every extension uninstall itself.
// [[spec/design_output/extension#a-file-another-program-owns]]

import { basename, dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { clock } from "../doors/clock.js";
import { disk } from "../doors/disk.js";

export const LIST = "extensions.json";
export const KEPT = "extensions.json.before-quackitect";

// [[spec/design_output/extension#a-file-another-program-owns]]
export function readEntries(text) {
  let said;
  try {
    said = JSON.parse(String(text ?? ""));
  } catch {
    return { entries: [], unreadable: true, dropped: 0, unwrapped: 0 };
  }

  const out = { entries: [], unreadable: false, dropped: 0, unwrapped: 0 };
  collect(said, out);
  return out;
}

function collect(node, out) {
  if (Array.isArray(node)) {
    for (const one of node) collect(one, out);
    return;
  }
  if (node && typeof node === "object") {
    if (typeof node.identifier?.id === "string" && node.identifier.id !== "") {
      out.entries.push(node);
      return;
    }
    if (Array.isArray(node.value)) {
      out.unwrapped += 1;
      collect(node.value, out);
      return;
    }
  }
  out.dropped += 1;
}

export function entryFor(id, version, dest, at) {
  return {
    identifier: { id },
    version,
    location: { $mid: 1, path: dest, scheme: "file" },
    relativeLocation: basename(dest),
    metadata: { installedTimestamp: at, source: "vsix" },
  };
}

// [[spec/design_output/extension#a-lost-id-stands-refused]]
export function upsert(said, mine) {
  const id = mine.identifier.id;
  const was = said.entries.map((one) => one.identifier.id);
  const kept = said.entries.filter((one) => one.identifier.id !== id);
  const entries = [...kept, mine];

  const holds = new Set(entries.map((one) => one.identifier.id));
  const lost = was.filter((one) => one !== id && !holds.has(one));
  return { entries, lost, replaced: was.includes(id) };
}

// [[spec/design_output/extension#a-file-another-program-owns]]
export function register(files, folder, mine) {
  const where = join(folder, LIST);
  const text = files.exists(where) ? files.read(where) : "[]";
  const said = readEntries(text);
  if (said.unreadable) {
    return { wrote: false, why: "the list reads as no JSON at all, so it stands as it is" };
  }

  const found = upsert(said, mine);
  if (found.lost.length) {
    return { wrote: false, why: `writing would lose ${found.lost.join(", ")}, so nothing went in` };
  }

  if (files.exists(where)) files.write(join(folder, KEPT), text);
  files.write(where, `${JSON.stringify(found.entries)}\n`);
  return { wrote: true, why: found.replaced ? "the entry stood already, and it stands again" : "the entry went in" };
}

// [[spec/design_output/extension#the-link-stands]]
export function linkedAt(files, dest, source) {
  if (!files.isLink(dest)) return false;
  return same(files.realOf(dest), files.realOf(source));
}

export function linkAt(files, dest, source) {
  if (!shape(dest).includes("/.vscode/extensions/")) {
    return { linked: false, why: `${dest} stands outside the editor's folder, so nothing goes there` };
  }
  if (linkedAt(files, dest, source)) return { linked: true, why: "the link stands already" };
  if (files.exists(dest)) files.remove(dest);
  files.link(source, dest);
  return { linked: true, why: "the link went in" };
}

export function registered(files, folder, id) {
  const where = join(folder, LIST);
  if (!files.exists(where)) return false;
  return readEntries(files.read(where)).entries.some((one) => one.identifier.id === id);
}

function shape(path) {
  return String(path).split("\\").join("/");
}

function same(one, two) {
  return shape(one).toLowerCase() === shape(two).toLowerCase();
}

export function rootHere() {
  return dirname(dirname(dirname(fileURLToPath(import.meta.url))));
}

export function manifestPath(root) {
  return join(root, "src", "extension", "package.json");
}

// [[spec/design_output/extension#a-box-names-its-home]]
export function homeIn(env) {
  return env.USERPROFILE || env.HOME || "";
}

// [[spec/design_output/extension#the-link-stands]]
function main(env, verb) {
  const home = homeIn(env);
  if (!home) {
    console.error("This box names no home folder, so the editor's list has none.");
    return 1;
  }

  const files = disk();
  const root = rootHere();
  const said = JSON.parse(files.read(manifestPath(root)));
  const id = `${said.publisher}.${said.name}`;
  const folder = join(home, ".vscode", "extensions");
  const dest = join(folder, `${id}-${said.version}`);
  const source = dirname(manifestPath(root));

  if (verb === "linked") {
    return linkedAt(files, dest, source) && registered(files, folder, id) ? 0 : 1;
  }

  const linked = linkAt(files, dest, source);
  console.log(`${id}: ${linked.why}.`);
  if (!linked.linked) return 1;
  const found = register(files, folder, entryFor(id, said.version, dest, clock().now().getTime()));
  console.log(`${id}: ${found.why}.`);
  return found.wrote ? 0 : 1;
}

if (process.argv[1]?.endsWith("editor.js")) process.exit(main(process.env, process.argv[2]));
